File: ta4j-core/src/test/java/org/ta4j/core/MarketEventTestContext.java
package org.ta4j.core;

import static org.assertj.core.api.Assertions.assertThat;
import static org.ta4j.core.TestUtils.GENERAL_OFFSET;
import static org.ta4j.core.TestUtils.assertNumEquals;

import java.time.Duration;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.UUID;
import java.util.stream.Stream;

import lombok.Getter;
import org.assertj.core.data.Offset;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.ta4j.core.backtest.BacktestBarSeries;
import org.ta4j.core.backtest.BacktestBarSeriesBuilder;
import org.ta4j.core.events.CandleReceived;
import org.ta4j.core.events.MarketEvent;
import org.ta4j.core.indicators.Indicator;
import org.ta4j.core.indicators.IndicatorContext;
import org.ta4j.core.indicators.bool.BooleanIndicator;
import org.ta4j.core.indicators.numeric.NumericIndicator;
import org.ta4j.core.mocks.MockMarketEventBuilder;
import org.ta4j.core.num.NaN;
import org.ta4j.core.num.Num;
import org.ta4j.core.num.NumFactory;

/**
 * This class simplifies testing by connecting market events and bar series through indicator context.
 *
 * Events are replayed to bar series that refreshes indicator context by created bar.
 */
public class MarketEventTestContext {
  private static final Logger log = LoggerFactory.getLogger(MarketEventTestContext.class);

  private Queue<MarketEvent> marketEvents;
  private final IndicatorContext indicatorContext = IndicatorContext.empty();
  @Getter
  private BacktestBarSeries barSeries =
      new BacktestBarSeriesBuilder().withIndicatorContext(this.indicatorContext).build();
  private Duration candleDuration = Duration.ofDays(1);
  private Instant startTime = Instant.EPOCH;


  /**
   * Sets OHLC to the same value.
   *
   * @param prices to set for each candle
   *
   * @return this
   */
  public MarketEventTestContext withCandlePrices(final double... prices) {
    this.marketEvents = new LinkedList<>(
        new MockMarketEventBuilder()
            .withStartTime(this.startTime)
            .withCandleDuration(this.candleDuration)
            .withCandlePrices(prices)
            .build()
    );
    return this;
  }


  public MarketEventTestContext withMarketEvents(final List<MarketEvent> marketEvents) {
    this.marketEvents = new LinkedList<>(marketEvents);
    return this;
  }


  public MarketEventTestContext withDefaultMarketEvents() {
    this.marketEvents = new LinkedList<>(new MockMarketEventBuilder().withDefaultData().build());
    return this;
  }


  public MarketEventTestContext withIndicator(final Indicator<?> indicator) {
    this.indicatorContext.add(indicator, UUID.randomUUID().toString());
    return this;
  }


  public MarketEventTestContext withIndicator(final Indicator<?> indicator, final String name) {
    this.indicatorContext.add(indicator, name);
    return this;
  }


  public MarketEventTestContext withIndicators(final Indicator<?>... indicator) {
    Stream.of(indicator).forEach(this::withIndicator);
    return this;
  }


  /**
   * This method should be called before creation of any indicator to preserve precision.
   *
   * @param factory that wil build {@link Num}
   *
   * @return this
   */
  public MarketEventTestContext withNumFactory(final NumFactory factory) {
    this.barSeries =
        new BacktestBarSeriesBuilder()
            .withNumFactory(factory)
            .withIndicatorContext(this.indicatorContext)
            .build();
    return this;
  }


  public boolean advance() {
    log.trace("\t ### advance");

    final var marketEvent = this.marketEvents.poll();
    if (marketEvent == null) {
      return false;
    }

    switch (marketEvent) {
      case final CandleReceived e -> this.barSeries.onCandle(e);
      default -> throw new IllegalStateException("Unexpected value: " + marketEvent);
    }

    return true;
  }


  public NumericIndicator getFirstNumericIndicator() {
    return (NumericIndicator) this.indicatorContext.getFirst();
  }


  public BooleanIndicator getFisrtBooleanIndicator() {
    return (BooleanIndicator) this.indicatorContext.getFirst();
  }


  public MarketEventTestContext fastForwardUntilStable() {
    while (this.indicatorContext.isNotEmpty() && !this.indicatorContext.isStable()) {
      fastForward(1);
    }

    return this;
  }


  public static void assertNextNaN(final MarketEventTestContext context, final NumericIndicator indicator) {
    context.advance();
    assertNumEquals(NaN.NaN, indicator.getValue());
  }


  public static void assertNextFalse(final MarketEventTestContext context) {
    context.advance();
    assertThat(context.getFisrtBooleanIndicator().getValue()).isFalse();
  }


  public static void assertNextTrue(final MarketEventTestContext context) {
    context.advance();
    assertThat(context.getFisrtBooleanIndicator().getValue()).isTrue();
  }


  /**
   * @param bars how many bars should be skipped.
   *
   * @return this
   */
  public MarketEventTestContext fastForward(final int bars) {
    log.debug("Fast forward =====> {}", bars);
    for (int i = 0; i < bars; i++) {
      if (!advance()) {
        throw new IllegalStateException("Fast forward failed at index " + i);
      }
    }
    return this;
  }


  public MarketEventTestContext assertCurrent(final double expected) {
    assertCurrent(getFirstNumericIndicator(), expected);
    return this;
  }


  private MarketEventTestContext assertCurrent(final NumericIndicator indicator, final double expected) {
    assertNumEquals(expected, indicator.getValue());
    return this;
  }


  private NumericIndicator getNumericIndicator(final String indicatorName) {
    return (NumericIndicator) this.indicatorContext.get(indicatorName);
  }


  public MarketEventTestContext assertNext(final double expected) {
    assertNext(getFirstNumericIndicator(), expected);
    return this;
  }


  private MarketEventTestContext assertNext(final NumericIndicator indicator, final double expected) {
    if (!advance()) {
      throw new IllegalStateException("Next failed");
    }

    assertNumEquals(expected, indicator.getValue());
    return this;
  }


  public MarketEventTestContext assertIndicatorEquals(
      final Indicator<Num> expectedIndicator,
      final Indicator<Num> indicator
  ) {
    while (advance()) {
      assertThat(indicator.getValue().doubleValue())
          .isCloseTo(
              expectedIndicator.getValue().doubleValue(),
              Offset.offset(GENERAL_OFFSET)
          );
    }

    return this;
  }


  public IndicatorAsserts onIndicator(final String name) {
    return new IndicatorAsserts(name);
  }


  /**
   * Adaptation of market events to executed position testing.
   *
   * There is some magic due to different clocks. This method adjusts clock to match with market events.
   *
   * Some criteria are not event based, but are retrospective instead. So for them, we have to replay all events
   * and calculate them on historical data.
   */
  public TradingRecordTestContext toTradingRecordContext() {
    return new TradingRecordTestContext(this);
  }


  public void withBarListener(final BarListener barListener) {
    this.barSeries.clearListeners();
    this.barSeries.addListener(barListener);
  }


  public MarketEventTestContext withCandleDuration(final ChronoUnit candleDuration) {
    this.candleDuration = candleDuration.getDuration();
    return this;
  }


  public MarketEventTestContext withStartTime(final Instant startTime) {
    this.startTime = startTime;
    return this;
  }


  public class IndicatorAsserts {
    private final String indicatorName;


    public IndicatorAsserts(final String indicatorName) {
      this.indicatorName = indicatorName;
      if (!MarketEventTestContext.this.indicatorContext.contains(indicatorName)) {
        throw new IllegalStateException("Indicator " + indicatorName + " not found");
      }
    }


    public IndicatorAsserts assertNext(final double expected) {
      MarketEventTestContext.this.assertNext(getNumericIndicator(this.indicatorName), expected);
      return this;
    }


    public IndicatorAsserts assertCurrent(final double expected) {
      MarketEventTestContext.this.assertCurrent(getNumericIndicator(this.indicatorName), expected);
      return this;
    }
  }
}

File: ta4j-core/src/test/java/org/ta4j/core/MarketEventTestContext.kt
package org.ta4j.core

import java.time.Duration
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.*
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.data.Offset
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.ta4j.core.TestUtils.GENERAL_OFFSET
import org.ta4j.core.TestUtils.assertNumEquals
import org.ta4j.core.api.Indicator
import org.ta4j.core.api.callback.BarListener
import org.ta4j.core.backtest.BacktestBarSeries
import org.ta4j.core.backtest.BacktestBarSeriesBuilder
import org.ta4j.core.events.CandleReceived
import org.ta4j.core.events.MarketEvent
import org.ta4j.core.indicators.IndicatorContext
import org.ta4j.core.indicators.IndicatorContext.IndicatorIdentification
import org.ta4j.core.indicators.TimeFrame
import org.ta4j.core.indicators.bool.BooleanIndicator
import org.ta4j.core.indicators.numeric.NumericIndicator
import org.ta4j.core.mocks.MockMarketEventBuilder
import org.ta4j.core.num.NaN
import org.ta4j.core.num.Num
import org.ta4j.core.num.NumFactory
import org.ta4j.core.utils.TimeFrameMapping

/**
 * This class simplifies testing by connecting market events and bar series through indicator context.
 *
 * Events are replayed to bar series that refreshes indicator context by created bar.
 */
class MarketEventTestContext {
    private lateinit var marketEvents: Queue<MarketEvent>
    private val indicatorContext: IndicatorContext = IndicatorContext.empty(TimeFrame.DAY)


    var barSeries: BacktestBarSeries =
        BacktestBarSeriesBuilder().withIndicatorContext(indicatorContext).build()
        private set
    private var candleDuration: Duration = TimeFrameMapping.getDuration(TimeFrame.DAY)
    private var startTime: Instant = Instant.EPOCH


    /**
     * Sets OHLC to the same value.
     *
     * @param prices to set for each candle
     *
     * @return this
     */
    fun withCandlePrices(vararg prices: Double): MarketEventTestContext = apply {
        marketEvents = LinkedList(
            MockMarketEventBuilder()
                .withStartTime(startTime)
                .withCandleDuration(candleDuration)
                .withCandlePrices(*prices)
                .build()
        )
    }

    fun withMarketEvents(marketEvents: List<MarketEvent>): MarketEventTestContext = apply {
        this.marketEvents = LinkedList(marketEvents)
    }


    fun withDefaultMarketEvents(): MarketEventTestContext = apply {
        marketEvents = LinkedList(MockMarketEventBuilder().withDefaultData().build())
    }


    fun withIndicator(indicator: Indicator<*>): MarketEventTestContext = apply {
        indicatorContext.add(indicator)
    }


    fun withIndicator(indicator: Indicator<*>, name: String): MarketEventTestContext = apply {
        indicatorContext.add(indicator, IndicatorIdentification(name))
    }


    fun withIndicators(vararg indicator: Indicator<*>): MarketEventTestContext = apply {
        indicator.forEach { withIndicator(it) }
    }


    /**
     * This method should be called before creation of any indicator to preserve precision.
     *
     * @param factory that wil build [Num]
     *
     * @return this
     */
    fun withNumFactory(factory: NumFactory): MarketEventTestContext = apply {
        barSeries =
            BacktestBarSeriesBuilder()
                .withNumFactory(factory)
                .withTimeFrame(barSeries.timeFrame)
                .withIndicatorContext(indicatorContext)
                .build()
    }

    /**
     * Sets a custom barSeries for this context.
     * This is useful when you need a specific timeframe that differs from the default DAY timeframe.
     */
    fun withBarSeries(customBarSeries: BacktestBarSeries): MarketEventTestContext = apply {
        barSeries = customBarSeries
    }
    


    fun advance(): Boolean {
        log.trace("\t ### advance")

        val marketEvent = marketEvents.poll()
        if (marketEvent == null) {
            return false
        }

        when (marketEvent) {
            is CandleReceived -> barSeries.onCandle(marketEvent)
            else -> error("Unexpected value: $marketEvent")
        }

        return true
    }


    val firstNumericIndicator: NumericIndicator?
        get() = indicatorContext.first as NumericIndicator?


    val fisrtBooleanIndicator: BooleanIndicator?
        get() = indicatorContext.first as BooleanIndicator?


    fun fastForwardUntilStable(): MarketEventTestContext = apply {
        while (indicatorContext.isNotEmpty && !indicatorContext.isStable) {
            fastForward(1)
        }
    }


    fun assertNextNaN(): MarketEventTestContext = apply {
        fastForward(1)
        assertNumEquals(NaN, firstNumericIndicator!!.value)
    }

    fun assertCurrentNaN(): MarketEventTestContext = apply {
        assertNumEquals(NaN, firstNumericIndicator!!.value)
    }


    fun assertNextFalse(): MarketEventTestContext = apply {
        fastForward(1)
        assertThat(fisrtBooleanIndicator!!.value).isFalse()
    }


    fun assertNextTrue(): MarketEventTestContext = apply {
        fastForward(1)
        assertThat(fisrtBooleanIndicator!!.value).isTrue()
    }


    /**
     * @param bars how many bars should be skipped.
     *
     * @return this
     */
    fun fastForward(bars: Int): MarketEventTestContext = apply {
        log.debug("Fast forward =====> {}", bars)
        for (i in 0..<bars) {
            check(advance()) { "Fast forward failed at index " + i }
        }
    }


    fun assertCurrent(expected: Double): MarketEventTestContext = apply {
        assertCurrent(firstNumericIndicator!!, expected)
    }


    private fun assertCurrent(indicator: NumericIndicator, expected: Double): MarketEventTestContext = apply {
        assertNumEquals(expected, indicator.value)
    }


    private fun getNumericIndicator(indicatorName: IndicatorIdentification): NumericIndicator? {
        return indicatorContext.getNumericIndicator(indicatorName)
    }


    fun assertNext(expected: Double): MarketEventTestContext = apply {
        assertNext(firstNumericIndicator!!, expected)
    }


    private fun assertNext(indicator: NumericIndicator, expected: Double): MarketEventTestContext = apply {
        check(advance()) { "Next failed" }

        assertNumEquals(expected, indicator.value)
    }


    fun assertIndicatorEquals(
        expectedIndicator: Indicator<Num>,
        indicator: Indicator<Num>,
    ): MarketEventTestContext = apply {
        while (advance()) {
            assertThat(indicator.value.doubleValue())
                .describedAs { "Failed at index ${barSeries.currentIndex} and time ${barSeries.currentTime}" }
                .isCloseTo(
                    expectedIndicator.value.doubleValue(),
                    Offset.offset(GENERAL_OFFSET)
                )
        }
    }


    fun onIndicator(name: String): IndicatorAsserts = IndicatorAsserts(name)


    /**
     * Adaptation of market events to executed position testing.
     *
     * There is some magic due to different clocks. This method adjusts clock to match with market events.
     *
     * Some criteria are not event based, but are retrospective instead. So for them, we have to replay all events
     * and calculate them on historical data.
     */
    fun toTradingRecordContext(): TradingRecordTestContext = TradingRecordTestContext(this)


    fun withBarListener(barListener: BarListener) {
        barSeries.clearListeners()
        barSeries.addBarListener(barListener)
    }


    fun withCandleDuration(candleDuration: ChronoUnit): MarketEventTestContext = apply {
        this.candleDuration = candleDuration.duration
    }


    fun withStartTime(startTime: Instant): MarketEventTestContext = apply {
        this.startTime = startTime
    }


    fun assertIsStable() {
        val indicator = getAnyIndicator()
        assertThat(indicator?.isStable).isTrue()
    }

    fun assertIsUnStable() {
        val indicator = getAnyIndicator()
        assertThat(indicator?.isStable).isFalse()
    }

    private fun getAnyIndicator() = firstNumericIndicator ?: fisrtBooleanIndicator


    inner class IndicatorAsserts(indicatorName: String) {
        private val indicatorId: IndicatorIdentification = IndicatorIdentification(indicatorName)


        init {
            check(indicatorId in this@MarketEventTestContext.indicatorContext) { "Indicator $indicatorName not found" }
        }


        fun assertNext(expected: Double): IndicatorAsserts = apply {
            this@MarketEventTestContext.assertNext(getNumericIndicator(indicatorId)!!, expected)
        }


        fun assertCurrent(expected: Double): IndicatorAsserts = apply {
            this@MarketEventTestContext.assertCurrent(getNumericIndicator(indicatorId)!!, expected)
        }
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(MarketEventTestContext::class.java)
    }
}
